[FSDP][Easy] Save unpadded/padded unsharded sizes as attributes

awgu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #84911 [FSDP] Add `use_orig_params`
* #84910 [FSDP] Post-bwd prefetch with reverse pre-forward order
* #83917 [FSDP] Add rate limiter
* **#84366 [FSDP][Easy] Save unpadded/padded unsharded sizes as attributes**
* #83665 [FSDP] Generalize prefetching; lower unshard/reshard to handle
* #84761 [FSDP][Easy] Minor cleanup
* #84601 [FSDP] Subtest prefetching for `test_fsdp_grad_acc.py`
* #84600 [FSDP] Remove `forward_prefetch`

Differential Revision: [D39331199](https://our.internmc.facebook.com/intern/diff/D39331199)

Commit:
[FSDP][Easy] Save unpadded/padded unsharded sizes as attributes

[ghstack-poisoned]

Commit:
Update on "[FSDP][Easy] Save unpadded/padded unsharded sizes as attributes"




[ghstack-poisoned]

Commit:
Update on "[FSDP][Easy] Save unpadded/padded unsharded sizes as attributes"




[ghstack-poisoned]

rohan-varma commented on torch/distributed/fsdp/flat_param.py:
```diff
@@ -128,8 +128,11 @@ class FlatParameter(nn.Parameter):
-            (without padding).
+        _unpadded_unsharded_size (torch.Size): Unsharded flattened parameter's
+            size without padding.
+        _padded_unsharded_size (torch.Size): Unsharded flattened parameter's
```
Might result in some confusing code in the future if these attributes are being used across FSDP instances where maybe one instance isn't sharded. 

Could we consider setting it for non-sharded strategies and just having it equal to unpadded_unsharded_size since we don't need to pad?

awgu replied:
I thought about this for a while, and I want to play devil's advocate.

As is, the sharded strategies are handled differently compared to `NO_SHARD` in a few places. It is only in those sharded strategy code paths that `_padded_unsharded_size` is used. I would find it a yellow flag if we are using `_padded_unsharded_size` for both `NO_SHARD` and the sharded strategies; it would be an indication that we may actually want to be using the `_unpadded_unsharded_size` or more broadly the unpadded unsharded flattened parameter.

An example of this would be the old `summon_full_params(offload_to_cpu=True)`. This previously used the *padded* unsharded flattened parameter, but since it was agnostic to whether the sharding strategy was a sharded one or not, we could get away with simply using the *unpadded* unsharded flattened parameter. In other words there is no need to include the padding. (This is what the "generalize prefetching" PR does.)

Given this, it may actually prevent silent errors/mistakes to have `_padded_unsharded_size` only be defined for sharded strategies and hence ensure that we do not inadvertently use the _padded_ unsharded flattened parameter when the _unpadded_ one suffices.

rohan-varma reviewed (approved):
LGTM

Commit:
Update on "[FSDP][Easy] Save unpadded/padded unsharded sizes as attributes"




[ghstack-poisoned]

Commit:
Update on "[FSDP][Easy] Save unpadded/padded unsharded sizes as attributes"




[ghstack-poisoned]

Commit:
Update on "[FSDP][Easy] Save unpadded/padded unsharded sizes as attributes"




[ghstack-poisoned]

Commit:
Update on "[FSDP][Easy] Save unpadded/padded unsharded sizes as attributes"




[ghstack-poisoned]

Commit:
Update on "[FSDP][Easy] Save unpadded/padded unsharded sizes as attributes"




[ghstack-poisoned]

Commit:
Update on "[FSDP][Easy] Save unpadded/padded unsharded sizes as attributes"




[ghstack-poisoned]

Commit:
Update on "[FSDP][Easy] Save unpadded/padded unsharded sizes as attributes"




[ghstack-poisoned]

Commit:
Update on "[FSDP][Easy] Save unpadded/padded unsharded sizes as attributes"




[ghstack-poisoned]

awgu commented:
@awgu has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D39331199).

Commit:
Update on "[FSDP][Easy] Save unpadded/padded unsharded sizes as attributes"


Differential Revision: [D39331199](https://our.internmc.facebook.com/intern/diff/D39331199)

[ghstack-poisoned]

Commit:
Update on "[FSDP][Easy] Save unpadded/padded unsharded sizes as attributes"


Differential Revision: [D39331199](https://our.internmc.facebook.com/intern/diff/D39331199)

[ghstack-poisoned]

Commit:
Update on "[FSDP][Easy] Save unpadded/padded unsharded sizes as attributes"


Differential Revision: [D39331199](https://our.internmc.facebook.com/intern/diff/D39331199)

[ghstack-poisoned]

awgu commented:
@awgu has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D39331199).

Commit:
Update on "[FSDP][Easy] Save unpadded/padded unsharded sizes as attributes"


Differential Revision: [D39331199](https://our.internmc.facebook.com/intern/diff/D39331199)

[ghstack-poisoned]

Commit:
Update on "[FSDP][Easy] Save unpadded/padded unsharded sizes as attributes"


Differential Revision: [D39331199](https://our.internmc.facebook.com/intern/diff/D39331199)

[ghstack-poisoned]

Commit:
Update on "[FSDP][Easy] Save unpadded/padded unsharded sizes as attributes"


Differential Revision: [D39331199](https://our.internmc.facebook.com/intern/diff/D39331199)

[ghstack-poisoned]

Commit:
Update on "[FSDP][Easy] Save unpadded/padded unsharded sizes as attributes"


Differential Revision: [D39331199](https://our.internmc.facebook.com/intern/diff/D39331199)

[ghstack-poisoned]

Commit:
Update on "[FSDP][Easy] Save unpadded/padded unsharded sizes as attributes"


Differential Revision: [D39331199](https://our.internmc.facebook.com/intern/diff/D39331199)

[ghstack-poisoned]

Commit:
Update on "[FSDP][Easy] Save unpadded/padded unsharded sizes as attributes"


Differential Revision: [D39331199](https://our.internmc.facebook.com/intern/diff/D39331199)

[ghstack-poisoned]

Commit:
Update on "[FSDP][Easy] Save unpadded/padded unsharded sizes as attributes"


Differential Revision: [D39331199](https://our.internmc.facebook.com/intern/diff/D39331199)

[ghstack-poisoned]

awgu commented:
@awgu has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D39331199).

Commit:
Update on "[FSDP][Easy] Save unpadded/padded unsharded sizes as attributes"


Differential Revision: [D39331199](https://our.internmc.facebook.com/intern/diff/D39331199)

[ghstack-poisoned]

awgu commented:
@awgu has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D39331199).

Commit:
Update on "[FSDP][Easy] Save unpadded/padded unsharded sizes as attributes"


Differential Revision: [D39331199](https://our.internmc.facebook.com/intern/diff/D39331199)

[ghstack-poisoned]

Commit:
Update on "[FSDP][Easy] Save unpadded/padded unsharded sizes as attributes"


Differential Revision: [D39331199](https://our.internmc.facebook.com/intern/diff/D39331199)

[ghstack-poisoned]

awgu commented:
@awgu has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D39331199).

Commit:
Update on "[FSDP][Easy] Save unpadded/padded unsharded sizes as attributes"


Differential Revision: [D39331199](https://our.internmc.facebook.com/intern/diff/D39331199)

[ghstack-poisoned]

awgu commented:
@awgu has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D39331199).

Commit:
Update on "[FSDP][Easy] Save unpadded/padded unsharded sizes as attributes"


Differential Revision: [D39331199](https://our.internmc.facebook.com/intern/diff/D39331199)

[ghstack-poisoned]

awgu commented:
@awgu has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D39331199).

awgu commented:
@pytorchbot merge
